### ISSUE TITLE
Example in documentation for UpgradeModule 'Upgrading an AngularJS service' is wrong; results in "Unhandled Promise rejection: Cannot read property 'get' of undefined"

### ISSUE THREAD

**Issue by andrewl-telnyx** [angular/angular#18712]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

[Documentation example for UpgradeModule](https://angular.io/api/upgrade/static/UpgradeModule#example) does not adequately address issues in which angularJS's $injector being undefined.
This is directly related to: [This stackoverflow problem](https://stackoverflow.com/questions/45644207/angular-ngupgrade-inject-angularjs-service-into-angular-service-getting-unhand)
and [this open issue](https://github.com/angular/angular/issues/14993) 

TLDR: trying to use upgrade module to inject an AngularJS service into an Angular service fails with "Cannot read property 'get' of undefined ; Zone: ...
Following the docs exactly results in the error. A hack that I found that's super gross ( don't judge me ) is to:

```
  useFactory: () => {
        return new Promise((resolve) => {
          setTimeout(function(){
            resolve(angular.element(document)
                .injector().get('the angularJS service'))
          },1);
        })
      },
```
There has to be a better way- and whatever that better way is, it NEEDS to be in the docs.

## Expected behavior
- You should be able to inject an AngularJS service into an Angular service and it should work.
- Something should be done so that $injector is not undefined (and if it's undefined for this reason, it'd be cool if it pointed you to a solution like how they did in angularJS).
- The docs for UpgradeModule need to be updated. 

## Minimal reproduction of the problem with instructions
[Here's a plnkr that uses the exact code from the docs...or at least as close as I could get it](http://plnkr.co/edit/QCj85Xsm3tkzHDbOL1fA?p=preview)

[Here's a description of how I ran into the problem](https://stackoverflow.com/questions/45644207/angular-ngupgrade-inject-angularjs-service-into-angular-service-getting-unhand) and more context. And also my hack in context.

## What is the motivation / use case for changing the behavior?
The NgUpgrade docs make it seem like this is going to be simple, and it, like most things in life, almost never is. Documentation is hard to write for sure, and all things considered the Angular team does as a decent job at it, but... with some of this stuff, you need more examples. And as complicated as it may be, it'd be way more helpful if the examples were in a plnkr or something.

That being said, the main motivation / use case for this is that I need to inject an AngularJS service into an Angular service.

## Environment

<pre><code>
Angular version: 4.2.4

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  Mac

Others:
I believe in you.
</code></pre>

**Comment by andrewl-telnyx** [comment]
(i wanted to check 2 boxes in the "i'm submitting a" section, but the comment said to only select one)

**Comment by gkalpak** [comment]
As the example in the doc shows, you should not specify `bootstrap` components when using `ngUpgrade`, because you don't want a component to be bootstrapped until both Angular and AngularJS are initialized. Instead, you should downgrade the component and let `ngUpgrade` do the bootstrapping. 

This [updated plnkr][1] shows a warking vrsion of your example.

##
Another minor issue (which does not affect this simple usecase) is that you should be bootstrapping AngularJS in the `ngDoBootstrap()` method. We have updated this in most places in the docs; #18487 should take care of the remaining places.

##
Closing, since everything works as expected afaict.

[1]: http://plnkr.co/edit/FqUcwDSmx0CyaMhjZvmU?p=preview

**Comment by andrewl-telnyx** [comment]
@gkalpak At Google you all appreciate constructive criticism, transparency, and curiosity at every level of the organization, so allow me please to provide this constructive criticism.

The two things you did to get my example to work were
1. you moved the 
```
angular
.module('ng1AppModule', [])
.directive('heroDetail', downgradeComponent({component: HeroDetailComponent}))
.factory('titleCase',
  () => (value: string) => value.replace(/((^|\s)[a-z])/g, (_, c) => c.toUpperCase())
);
```

AFTER my angular 2 stuff, but before the platformBrowserDynamic stuff.

2. In the angularJS app, you downgraded the HeroDetailComponent.

I implore you, look at the [UpgradeModule docs](https://angular.io/api/upgrade/static/UpgradeModule#example), and you will see that neither of those two things that you did are in there.
Look at the [Angular Upgrade section of the docs](https://angular.io/guide/upgrade) and you will see that the two things you specified to to do are not in there either.

You say it's in the docs. 
- Please, show me where in the docs for UpgradeModule and/or where in the upgrading section of the docs it says:
"you **should not specify bootstrap components when using ngUpgrade**, because you don't want a component to be bootstrapped until both Angular and AngularJS are initialized. Instead, you should downgrade the component and let ngUpgrade do the bootstrapping."
( I'm not being rhetorical here, I really want to know where it says those things, and where the two things you did are shown in any of the upgrade examples so that I can refer others to them later. )

***

The most frustrating thing about upgrading is that, if you do exactly what the docs show/tell you to do, in each section that the docs do it, the examples straight up don't work. There is somehow simultaneously too much information and not enough information (code). The code examples are all broken up into pieces, which makes sense for the explanations of things, but at the end of the document there's no "This is what all of what that code combined should look like. Look! See in this jsfiddle embedded in the page how this example works?"
I feel that, given the complexity of upgrading AngularJS to Angular, there should be working examples in the documentation.

That all being said, documentation is HARD to get right. You all are doing a great job of explaining the details of how the stuff works but the practical examples, in this case, appear to be falling short. Thank you for taking the time to respond. Please consider adding working examples to the docs; and you or someone may want to answer the stackoverflow question because that's probably a place where people are going to go to find answers to this problem until the docs are updated.

**Comment by gkalpak** [comment]
> I implore you, look at the UpgradeModule docs, and you will see that neither of those two things that you did are in there.

> 1. you moved the ...  AFTER my angular 2 stuff, but before the platformBrowserDynamic stuff.

I did that, because in order to downgrade a component you need to reference its class. And since classes aren't hoisted, the class declaration needs to come first. This is not Angular-specific; it is "plain old" ES2015. 

There was no need to move everything, I would have just moved the `downgradeComponent()` part, but I thought it is better to group them together.

> 2. In the angularJS app, you downgraded the HeroDetailComponent.

The docs mentions that _"AngularJS is always bootstrapped first and owns the root component"_ , which implies that in order to "enter" the Angular world, you need to go through a downgraded component.

This particular API docs page focuses on the `UpgradeModule` API and has limited info on how to create a whole hybrid app. The linked ngUpgrade guide should be much more useful a resource for that.

##
> You say it's in the docs.

All I said was that _"as the example in the doc shows, you should not specify bootstrap components when using ngUpgrade"_. And indeed the example in the docs shows no `bootstrap` property on the `NgModule` decorator.

The guide is more explicit:

> Note that you do not add a `bootstrap` declaration to the `@NgModule` decorator, since AngularJS will own the root template of the application.

##
> So basically I have to downgrade the component in then AngularJS app, even though I'm already specifying its' existence in the Angular app in order to use the AngularJS service in an Angular service. Do you think that is at all redundant or confusing- especially given that i have yet to see where in the documentation it says to do that?

If all you are trying to do is use a servie, then you don't need to downgrade the component. Sorry, if I misunderstood - from your example it looked like you are trying to use the component. If you want to use a service you can simply use it.

But, as already mentioned, since AngularJS owns the root component, you can only use an Angular component by downgrading it.

I understand that it might be confusing why on framework needs to own the root component, but there are certain technical limitations that might not be obvious when looking from the outside.

##
> The most frustrating thing about upgrading is that, if you do exactly what the docs show/tell you to do, in each section that the docs do it, the examples straight up don't work.

It's probably me, but I still not convinced :grin: I understand that stitching a full hybrid app together from scattered API docs is not easy, but it isn't supposed to be (API docs have a different purpose) and that is why we have the guides, which should guide you step by step through the whole process. If that is not the case, we should definitely fix that (happily accepting pointers).

##
> The code examples are all broken up into pieces, which makes sense for the explanations of things, but at the end of the document there's no "This is what all of what that code combined should look like. 

I can definitely see how a link to a full example at the end of each ngUpgrade API doc page would be useful. 👍 on that.

##
> I feel that, given the complexity of upgrading AngularJS to Angular, there should be working examples in the documentation.

FWIW, the examples are extracted from [full working examples][1] (with e2e tests to make sure they don't break down the road). But again, 👍 on somehow making them available as live examples.
(I think this is something that is not supported atm for the examples used in API docs. Only for examples used in guides.)

@andrewl-telnyx, can you open a new issue about (somehow) linking to full ngUpgrade examples (either based on API docs or guide examples)?

In case it wasn't obvious from my previous comment, I know that the ngUpgrade docs are far from perfect. That said, it is sometimes difficult for us (working on ngUpgrade) to realize what a painpoint might be in real upgrade situations (because something that is obvious to us might not be obvious to app developers). So, any concrete suggestion on docs improvements is mostly welcome (even more so if it comes in the form of a PR 😛).

[1]: https://github.com/angular/angular/tree/master/packages/examples/upgrade/
